File: internal/models/infrastructure/sonarqube.go
package infrastructure

import "time"

type SonarQubeInstance struct {
	ID        uint      `gorm:"primarykey" json:"id"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
	Name      string    `gorm:"size:100;not null;uniqueIndex" json:"name"`
	BaseURL   string    `gorm:"size:500;not null" json:"base_url"`
	Token     string    `gorm:"size:500" json:"token"`
	IsDefault bool      `gorm:"default:false" json:"is_default"`
	Status    string    `gorm:"size:20;default:'active'" json:"status"`
}

func (SonarQubeInstance) TableName() string {
	return "sonarqube_instances"
}

type SonarQubeProjectBinding struct {
	ID                uint      `gorm:"primarykey" json:"id"`
	CreatedAt         time.Time `json:"created_at"`
	UpdatedAt         time.Time `json:"updated_at"`
	SonarQubeID       uint      `gorm:"not null;index" json:"sonarqube_id"`
	SonarProjectKey   string    `gorm:"size:200;not null" json:"sonar_project_key"`
	SonarProjectName  string    `gorm:"size:200" json:"sonar_project_name"`
	DevopsAppID       *uint     `json:"devops_app_id"`
	DevopsAppName     string    `gorm:"size:200" json:"devops_app_name"`
	QualityGateStatus string    `gorm:"size:20" json:"quality_gate_status"`
}

func (SonarQubeProjectBinding) TableName() string {
	return "sonarqube_project_bindings"
}

File: web/src/services/sonarqube.ts
import request from '@/services/api'

export interface SonarQubeInstance {
  id: number
  created_at?: string
  updated_at?: string
  name: string
  base_url: string
  token: string
  is_default: boolean
  status: string
}

export interface SonarQubeProjectBinding {
  id: number
  sonarqube_id: number
  sonar_project_key: string
  sonar_project_name: string
  devops_app_id?: number
  devops_app_name: string
  quality_gate_status: string
}

export interface SonarProject {
  key: string
  name: string
  qualifier: string
  visibility: string
}

export interface QualityGate {
  status: string
  conditions: { status: string; metricKey: string; comparator: string; errorThreshold: string; actualValue: string }[]
}

export interface SonarMeasure {
  metric: string
  value: string
}

export interface SonarIssue {
  key: string
  rule: string
  severity: string
  component: string
  line: number
  message: string
  status: string
  type: string
}

export const sonarqubeApi = {
  // 实例
  listInstances: () => request.get('/sonarqube/instances'),
  getInstance: (id: number) => request.get(`/sonarqube/instances/${id}`),
  createInstance: (data: Partial<SonarQubeInstance>) => request.post('/sonarqube/instances', data),
  updateInstance: (id: number, data: Partial<SonarQubeInstance>) => request.put(`/sonarqube/instances/${id}`, data),
  deleteInstance: (id: number) => request.delete(`/sonarqube/instances/${id}`),
  testConnection: (id: number) => request.post(`/sonarqube/instances/${id}/test`),

  // 绑定
  listBindings: (instanceId: number) => request.get(`/sonarqube/instances/${instanceId}/bindings`),
  createBinding: (instanceId: number, data: Partial<SonarQubeProjectBinding>) =>
    request.post(`/sonarqube/instances/${instanceId}/bindings`, data),
  updateBinding: (bindingId: number, data: Partial<SonarQubeProjectBinding>) =>
    request.put(`/sonarqube/bindings/${bindingId}`, data),
  deleteBinding: (bindingId: number) => request.delete(`/sonarqube/bindings/${bindingId}`),

  // SonarQube API 代理
  listProjects: (instanceId: number, page?: number, pageSize?: number) =>
    request.get(`/sonarqube/instances/${instanceId}/projects`, { params: { page, pageSize } }),
  getQualityGate: (instanceId: number, projectKey: string) =>
    request.get(`/sonarqube/instances/${instanceId}/quality-gate`, { params: { projectKey } }),
  getMeasures: (instanceId: number, projectKey: string) =>
    request.get(`/sonarqube/instances/${instanceId}/measures`, { params: { projectKey } }),
  getIssues: (instanceId: number, projectKey: string, page?: number, pageSize?: number, severities?: string) =>
    request.get(`/sonarqube/instances/${instanceId}/issues`, { params: { projectKey, page, pageSize, severities } }),
}
